torch/onnx/utils.py

```diff
@@ -1004,35 +1002,88 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
-                    #   prim::Loop(iter, cond, input_1, ..., input_n)
-                    #     block0(iter, input_1, ..., input_n)
+                    # The converted ONNX graph
+                    # %10 : Bool(device=cpu) = onnx::Constant[value={0}]()
+                    # %14 : Bool(device=cpu) = onnx::Equal(%13, %8)
+                    # %15 : Bool(requires_grad=0, device=cpu) = onnx::Constant[value={0}]()
+                    # %16 : Long(6, strides=[1], device=cpu) = onnx::Identity(%input.1)
+                    # %17 : Float(10, 15, strides=[15, 1], device=cpu) = onnx::Identity(%embedding_matrix.1)
                     #
-                    # For `If` node, there is nothing to copy.
-                    # For `Loop` node, copy metadata for `iter`, `input_1`, ..., `input_n`.
-                    for i, b_in in enumerate(b.inputs()):
-                        if i == 0 and i < len(inputs):
-                            b_in.setType(inputs[i].type())
-                        if i > 0 and (i + 1) < len(inputs):
-                            b_in.setType(inputs[i + 1].type())
-                    torch._C._jit_pass_onnx_block(b, new_block, operator_export_type, env)
-                new_op_outputs = torch._C._jit_pass_fixup_onnx_controlflow_node(new_node, opset_version)
-                # Process Loop and If after subblock is converted.
-                from torch.onnx.symbolic_helper import _onnx_shape_inference
-                if _onnx_shape_inference:
-                    torch._C._jit_pass_onnx_node_shape_type_inference(new_node, _params_dict, opset_version)
-                return new_op_outputs
+                    # env: 65:10, 21:14, 22:10, input.53:16, weight:17
+                    # embedding_matrix.1 and input.1 are prim::Param, so we need convert to ONNX Identity to do shape inference.
+                    input_flag = inputs[0].node()['value'].tolist()
+                    const_value = all(input_flag) if isinstance(input_flag, list) else bool(input_flag)
+                    block_idx = 0 if const_value else 1
+                    current_b = list(n.blocks())[block_idx]
+                    is_sub_block = True
+                    env = torch._C._jit_pass_onnx_block(current_b, block, operator_export_type, env,
+                                                        is_sub_block)
+                    if_output_list = list(n.outputs())
+                    current_b_list = list(current_b.outputs())
+
+                    final_b_list = []
+                    for idx in range(len(if_output_list)):
+                        if current_b_list[idx] not in env:
+                            raise RuntimeError("The sub block ATen output " + current_b_list[idx] + " is not in env")
+                        if "onnx" in env[current_b_list[idx]].node().kind():
+                            onnx_b = env[current_b_list[idx]]
+                        else:
+                            onnx_b = g.op("Identity", env[current_b_list[idx]])
```

Comment:
I think maybe you mentioned this before, could you remind me again why(for what case) this is needed?

Reply:
Good point. This Identity operator is no longer needed, so I just remove this.
Previously it is because I need make Prim::Param node gets shape inference. In current logic, every onnx operator will trigger shape inference while inserting, so it won't need that.